wiki/608483-BuildaTool.md

```diff
@@ -0,0 +1,123 @@
+of the design of the NLP Sandbox.
+
+## API (micro-)services vs programs
+
+Most developers are familiar with the development of command-line programs.
+Examples of programs include Bash scripts, Python scripts, Java executables,
+etc. The behavior of a program is usually defined by command line arguments
+and/or the use of a configuration file. One point that characterizes a program
+is that after completing its initialization (e.g., detection of GPUs) and
+primary task (e.g., processing data), the program shuts down.
+
+Another characteristic of *using* a program - from a user perspective - is that
+unless you use a container technology like Docker or Singularity, running a
+Python/Java/other program requires that you install a Python/Java/other
+environment beforehand.
+
+An Application Programming Interface (API) service is different from a program
+on the above two points. A service is a process that is continuously running and
+waiting for requests to process. For example, a web service keep listening to
+requests from a *clients* (your browser) and returns the requested HTML pages.
+Another example are API services that enables a client to create and manage data
+remotely using a standard interface (API). Relying on a standard and
+documentated API specification enables a service and its clients to be developed
+using different programming languages and frameworks.
+
+These two points are among the main reasons why the NLP Sandbox aims to develop
+NLP Tools as API services. API services are also inherently cloud-friendly as
+they can wait in the cloud (public or private servers) and are ready to answer
+requests anytime. Similarly to a web service that respond to million of requests
+from different users concurrently, an API service can process multiple requests
+in parallel without overhead for the developer of the service. And if your
+organization has the required resources, it is always possible to deploy
+multiple instances of the same service behind a load balancer that will take
+care of distributing the requests among the different instances based on their
+CPU load, for example.
+
+## OpenAPI specification
+
+Requiring NLP Sandbox Users to develop NLP Tools as API services instead of
+programs is not a decision that we made lightly. The risk would have been that
+users would have been brought to develop outside of their zone of comfort.
+However, besides the benefits of adopting a design based on the use of APIs, the
+development of API services is actually faster than developing a program that
+offer the same features. This achievement is made possible by adopting the
+[OpenAPI specification] and leveraging the tools developed by the [OpenAPI
+community].
+
+## The Gist of developing an NLP Sandbox Tool
+
+The development of an API service as documented by the NLP Sandbox is outlined
+below:
+
+1. Download the OpenAPI specification of an NLP Tool (JSON or YAML document).
```

Comment:
No need to do this.  provide them with a link that they can generate the NLP tool.  For instance: https://nlpsandbox.github.io/nlpsandbox-schemas/date-annotator/1.0.0/openapi.json

```
openapi-generator generate -g ... -o . --package-name ... -i https://nlpsandbox.github.io/nlpsandbox-schemas/date-annotator/1.0.0/openapi.json
```

Reply:
Nice! Good to know!